apps/voting/contracts/Voting.sol

```diff
@@ -519,40 +545,35 @@ contract Voting is IForwarder, AragonApp {
      * @dev Internal function to cast a vote or object to.
      * @dev It assumes that voter can support or object to the vote
      * @param _voteId The identifier of the vote
-     * @param _supports Whether the voter supports the vote or not
+     * @param _supports Whether the voter supports the vote
      * @param _voter The address of the voter
-     * @param _isDelegate Whether the voter is a delegate or not
+     * @param _isDelegate Whether the voter is a delegate
      */
-    function _vote(uint256 _voteId, bool _supports, address _voter, bool _isDelegate) internal {
+    function _vote(uint256 _voteId, bool _supports, address _voter, bool _isDelegate, uint256 _votingPower) internal {
         Vote storage vote_ = votes[_voteId];
-
-        // This could re-enter, though we can assume the governance token is not malicious
-        uint256 voterStake = token.balanceOfAt(_voter, vote_.snapshotBlock);
         VoterState state = vote_.voters[_voter];
 
         // If voter had previously voted, decrease count
-        if (state == VoterState.Yea || state == VoterState.DelegateYea) {
-            vote_.yea = vote_.yea.sub(voterStake);
-        } else if (state == VoterState.Nay || state == VoterState.DelegateNay) {
-            vote_.nay = vote_.nay.sub(voterStake);
+        if (state != VoterState.Absent) {
+            if (state == VoterState.Yea || state == VoterState.DelegateYea) {
+                vote_.yea = vote_.yea.sub(_votingPower);
+            } else { // Remaining states are Nay and DelegateNay
+                vote_.nay = vote_.nay.sub(_votingPower);
+            }
         }
 
         if (_supports) {
-            vote_.yea = vote_.yea.add(voterStake);
+            vote_.yea = vote_.yea.add(_votingPower);
             vote_.voters[_voter] = _isDelegate ? VoterState.DelegateYea : VoterState.Yea;
         } else {
-            vote_.nay = vote_.nay.add(voterStake);
+            vote_.nay = vote_.nay.add(_votingPower);
             vote_.voters[_voter] = _isDelegate ? VoterState.DelegateNay : VoterState.Nay;
         }
 
-        emit CastVote(_voteId, _voter, _supports, voterStake);
+        emit CastVote(_voteId, _voter, _supports, _votingPower);
 
         if (_getVotePhase(vote_) == VotePhase.Objection) {
```

Comment:
Consider passing the voting phase as a parameter instead of reading it from the store each time. It will decrease the gas costs when the delegate votes for many voters.

Reply:
Addressed in f1f6de1606a6eda1999a0e1cc0ea895857101f88.